File: Java/tutorial.java
//Autor: JaviMGG

//Esto es un comentario en una linea
/**
 * Esto es
 * un comentario
 * en varias
 * lineas
 */

/**
 * Estructura de los archivos de JAVA y ARGOTS:
 * ------------------------------------------------------------------------
 * Las carpetas que contienen diversos archivos JAVA se denominan "Paquetes"
 * Las hojas/archivos que tiene codigo JAVA se denominan "Clases"
 * Las clases contienen variables, "imports" (opcionales) y los denominados "Métodos"
 * Los métodos tienen la siguiente estructura: cabecera(Parametros OPCIONALES){CODIGO}
 * La cabecera tiene sus "ambitos de vida" (mas abajo está explicado)
 * ------------------------------------------------------------------------
 * IMPORTANTE: en JAVA, el método MAIN puede estar en cualquier posicion dentro de la "class", NO ES NECESARIO QUE sea el ultimo metodo de la clase
 * debido a que JAVA es un lenguaje compilado (un compilador revisa todo hasta que vea que está bien y lo traduce a lenguaje maquina...)
 */

    //##################################### Importar clases/elementos a esta clase ##############################################
    //import java.math.BigDecimal;
    //import java.net.Socket;
    //si queremos que se obtengan muchos mas elementos (y no uno concreto)
    //import java.net.*; //EL * obtendrá todos los elementos que se encuentren en java.net

    //esto se verá en la parte de "Leer de teclado y mostrarlo por pantalla"
    import java.util.Scanner;



public class tutorial {



    //##################################### Escribir HolaMundo (hacer un print) #################################################
    
    /**
     * Dos tipos de prints mas comunes:
     * print sin salto de linea: System.out.print("");
     * print con salto de linea: System.out.println(""); O System.out.print("...\n")
    */
    public static void main(String[] args) {
        System.out.print("holaMundo sin salto de linea");
        System.out.print("holaMundo con salto de linea");
    }



    //##################################### Tipo de datos, tipos de instancias y ambito de vida #################################

    /**
     * Ambito de vida:
     * ------------------------------------------------------------------------
     * Public       = TODOS PUEDEN VER Y UTILIZAR LA VARIABLE QUE TENGA ESTE "ambito" da igual si es esta CLASE O NO
     * Private      = Los metodos que "hereden" de este, NO PUEDEN USARLOS (tanto metodos como variables), es decir, si contiene PRIVATE, es EXCLUSIVO DE ESTA CLASE
     * Protected    = Los metodos que "hereden" de este, SI PUEDEN USARLOS (tanto metodos como variables), es decir, si contiene PROTECTED, ES EXCLUSIVO DE ESTA CLASE Y de sus "hijos"
     * 
     * Tipos de instancias:
     * ------------------------------------------------------------------------ 
     * Static       = Metodo de clase, no depende de un objeto.
     * Synchronized = Control de concurrencia, evita que varios hilos ejecuten el método a la vez sobre el mismo objeto.
     * (nada)       = Metodo de instancia normal.
     * Final        = UNICAMENTE SI EL VALOR ES CONSTANTE, ES DECIR, NO SE PODRÁ ALTERAR/MODIFICAR
     * 
     * Existen distinto tipos de datos:
     * ------------------------------------------------------------------------
     * Enteros      (int)
     * Reales       (double)
     * Float        (float)
     * Caracteres   (char)
     * Boolean      (boolean)
     * String       (String)
     * 
    */

    //ejemplo de posibles combinaciones de las variables y sus declaraciones:
    public int variableInt;                 //se puede declarar sin valor, pero luego hay que darle valor o DARA ERROR
    public double variableDouble = 2.44;    //variable declarada de tipo double
    public static boolean variableBooleana = true;
    private char variableCharacter = 'c';



    //##################################### Declarar y llamar una funcion ##############################################
    
    /**
     * Funciones/metodos: usados para poder ejecutar el codigo que tiene DENTRO
     * Tipos de funciones:
     * ------------------------------------------------------------------------
     * Enteros      (int)       => ES NECESARIO DEVOLVER ALGO/ HACE FALTA QUE LA ULTIMA LINEA DEL METODO SEA "return (tipo de dato correspondiente)"
     * Reales       (double)    => ES NECESARIO DEVOLVER ALGO/ HACE FALTA QUE LA ULTIMA LINEA DEL METODO SEA "return (tipo de dato correspondiente)"
     * Float        (float)     => ES NECESARIO DEVOLVER ALGO/ HACE FALTA QUE LA ULTIMA LINEA DEL METODO SEA "return (tipo de dato correspondiente)"
     * Caracteres   (char)      => ES NECESARIO DEVOLVER ALGO/ HACE FALTA QUE LA ULTIMA LINEA DEL METODO SEA "return (tipo de dato correspondiente)"
     * Boolean      (boolean)   => ES NECESARIO DEVOLVER ALGO/ HACE FALTA QUE LA ULTIMA LINEA DEL METODO SEA "return (tipo de dato correspondiente)"
     * String       (String)    => ES NECESARIO DEVOLVER ALGO/ HACE FALTA QUE LA ULTIMA LINEA DEL METODO SEA "return (tipo de dato correspondiente)"
     * void         (vacio)     => NO ES NECESARIO DEVOLVER NADA/ NO HACE FALTA "return"
    */
    public int getVariableInt(){
        return variableInt;          //SI LA VARIABLE en este caso NO ES ESTATICA Y EL METODO SI LO ES, DARÁ ERROR
    }

    public static boolean getVariableBooleana(){
        return variableBooleana;    //En este caso, tanto variable como metodo SON STATIC AMBOS
    }

    public static int suma(int x, int y){
        int z = x + y;
        return z;                   //se podia haber puesto directamente: return x + y
    }

    public void setCaracterNuevo(char x){
        variableCharacter = x;
    }

    public char getCaracterActual(){
        return variableCharacter;
    }

    public void ejemploLlamarUnaFuncion(){
        getVariableBooleana();              //llama a la funcion getVariableBooleana sin parametros
        int valor1 = 5;
        int valor2 = 8;
        int sumar = suma(valor1, valor2);    

        //llama a la funcion suma pasandole los valores 5 y 8 y ahora, la variable "sumar" será el resultado de lo que devuelva la funcion suma
        System.out.println("la suma de "+ valor1 + " y " + valor2 + " es: " + sumar);
    }



    //##################################### Condiciones ##############################################

    /**
     * Tipos de condiciones:
     * ------------------------------------------------------------------------
     * 1- if, else if, else     => if(condicion){cosa a ejecutar} else if(condicion "secundaria"){cosa a ejecutar} else{cosa a ejecutar si no se cumple las condiciones}
     * 2- switch                => switch(variable que determina los casos){case X: cosa a ejecutar si variable vale X...}
     * 3- operadores ternarios  => (comparacion) ? (si se cumple la comparacion) : (si no se cumple la comparacion)
     * 
    */
    public void ejemploCondiciones(){
        int x = 18;

        //if, else if, else
        if(x == 18){
            System.out.println("tienes 18 años");
        }else if (x < 18){
            System.out.println("tienes menos de 18 años");
        }
        else{
            System.out.println("tienes más de 18 años");
        }
        /**
         * IMPORTANTE, SI el if, else if, else tiene SOLAMENTE UNA INSTRUCCION A EJECUTAR (por ejemplo un print), SE PUEDE ELIMINAR EL {}:
         * if(x == 18)
         *      System.out.println("tienes 18 años");
         * else if (x < 18)
         *      System.out.println("tienes menos de 18 años");
         * else
         *      System.out.println("tienes más de 18 años");
         * 
        */

        //Switch
        x = 17;
        switch (x) {
            case 16 -> {System.out.println("x vale 16");}       //si x vale 16
            case 18 -> {System.out.println("x vale 18");}       //si x vale 18
            default -> System.out.println("x NO TIENE VALOR");  //si x no es ni 16 ni 18
        }

        //otra forma de declarar un switch
        switch (x) {
            case 17:                                            //si x vale 17
                System.out.println("x vale 17");
                break;                                          //si no existiera el break, se ejecutarian los dos "case"
            case 12:                                            //si x vale 12
                System.out.println("x vale 12");
                break;
            default:                                            //si x no tiene un "valor declarado"
                System.out.println("x no tiene valor, es un valor por defecto");
                //al ser el "valor por defecto", NO ES NECESARIO EL BREAK
        }

        //Operadores ternarios
        /**
         * Similar a un if, else
         * estructura:
         * (condicion a cumplir) ? (codigo a ejecutar si se cumple la condicion) : (codigo a ejecutar no se cumple la condicion)
        */
        String s = (variableBooleana) ? ("hola") : ("adios");
        System.out.print(s);
    }



    //##################################### Bucles ##############################################
    
    /**
     * Tres tipos de bucles:
     * ------------------------------------------------------------------------
     * for                      => for(desde donde comienzo; cuando pararé; como avanzo)
     * while                    => desde donde comienzo; while(condicion de parada){... como avanzo}
     * do while                 => desde donde comienzo; do {...como avanzo}while(condicion de parada) 
     * ------------------------------------------------------------------------
     * IMPORTANTE: 
     * el bucle do{}while(); SI QUE SE EJECUTA AL MENOS UNA VEZ, 
     * EL bucle while(){} SI INICIALMENTE NO CUMPLE LA CONDICION, NO SE EJECUTA, el do{}while(); SI
    */
    public static void ejemploBucles(){

        //bucle for
        for (int i = 0; i < 10; i++) { //en este ejemplo se lee asi: i = 0, i se ira incrementando en 1 hasta que sea menor que 10 y en cada iteracion, se ejecutara el codigo situado entre {}
            System.out.println(i);
        }

        //bucle while
        int i = 0;
        while(i < 10){
            System.out.println(i);
            i++;
        }

        //bucle do while
        i = 0;
        do{
            System.out.println(i);
            i++;
        }while(i < 10);
    }

    //##################################### Bloque try catch ##############################################
    
    /**
     * Usado para capturar excepciones.
     * Si por ejemplo tenemos una funcion que durante el proceso puede provocar el error, al saltar a dicho error,
     * se captura y se ejecuta el bloque catch
    */
    public static void ejemploTryCatch(int [] a){
        System.out.println("Array actual"+ a.toString());
        try {
            int suma = a[0];
            for (int i = 1; i < a.length; i++) {
                System.out.println(suma+=a[i+1]); //si I vale como máximo 5 y acaba valiendo 6, al acceder a un elemento que no existe (fuera del indice), salta excepcion
            }
        } catch (IndexOutOfBoundsException e) {
            System.out.println("VALOR FUERA DE INDICE"); //al capturarse la excepcion "e", se mostrará por pantalla lo del println
        }
    }



    //##################################### Leer de teclado y mostrarlo por pantalla ######################

    //Para ello usaremos la libreria de java.util.scanner que se encuentra arriba al inicio
    public static void ejemploScanner(){

        Scanner s = new Scanner(System.in);
        String palabraEscrita = s.nextLine();

        System.out.println(palabraEscrita);
    }

} //lo que haya debajo de esta linea NO PERTENECERÁ A LA CLASE Y DARÁ ERROR

File: C/tutorial.c
// Autor: JaviMGG

// Esto es un comentario de línea en C.

/**
 * Esto es un comentario
 * de bloque
 * en C.
*/



// ################################## Importar librerias ###########################################
#include <stdio.h>    //Libreria estándar de entrada y salida
#include <stdlib.h>   //Libreria estándar de utilidades
#include <string.h>   //Libreria para manejo de cadenas
// #include <mpi.h>   //Libreria para programación paralela con MPI
// #include <omp.h>   //Libreria para programación paralela con OpenMP
#include <stdbool.h>  //Libreria para tipo de dato booleano



// ################################# Tipos de datos ################################################
// estructura: <tipo de dato> nombre = <valor(opcional)>;

// Tipos de datos primitivos en C
int entero = 10;        // Entero
float flotante = 10.5f; // Número de punto flotante
double doble = 20.99;   // Doble precisión
char caracter = 'A';    // Carácter
double valorvacio;      // No tiene valor asignado, se le puede asignar uno posteriormente
bool booleano = 1;      // Booleano (0 o 1) // Requiere #include <stdbool.h>

// Cadenas de caracteres
char cadena[] = "Hola, Mundo!"; // Cadena de caracteres
int valores[];
double matriz[3][3];        // Matriz de 3x3

// Valores constantes
const float PI = 3.14159;   // Constante de tipo flotante
const int MAX_VALOR = 100;  // Constante de tipo entero
#define MAX_TAMANO 50       // Constante definida con #define



// ################################# Valores static ################################################
// lo que tenga "static" previamente, UNICAMENTE puede ser accedido desde este archivo, NO desde cualquier parte

static void funcionConStatic()
{
    static int variable = 0;
}



// ################################# Mostrar por pantalla valores ##################################
int entero = 0;
long numeroLong = 234567654322;
float flotante = 0.0f;
double doble = 0.0;
char caracter = ' ';
char cadena[100] = ""; // Cadena de caracteres con espacio para 100 caracteres
bool booleano = false; // Valor booleano inicializado a false
enum diasSemana {      // tipo de dato con un conjunto de constantes nombradas
    Lunes,
    Martes,
    Miercoles,
    Jueves,
    Viernes,
    Sabado,
    Domingo
};

void mostrarValores() 
{
    printf("Valor del entero: %d\n", entero);
    printf("Valor del flotante: %.2f\n", flotante);
    printf("Valor del doble: %.2lf\n", doble);
    printf("Valor del caracter: %c\n", caracter);
    printf("Valor de la cadena: %s\n", cadena);
    printf("Valor del booleano: %d\n", booleano); // Imprime 0 o 1 dependiendo del valor booleano
}



// ################################# Bucles ########################################################
/**
 * Estructura del bucle for:
 * for (inicialización; condición; incremento) {
 *     // Código a ejecutar en cada iteración
 * }
 * 
 * 
 * Estructura del bucle while:
 * while (condición) {
 *     // Código a ejecutar mientras la condición sea verdadera
 * }
 * 
 * 
 * Estructura del bucle do-while:
 * do {
 *     // Código a ejecutar al menos una vez
 * } while (condición);
*/

void mostrarBucles()
{
    int count = 10;
    for (int i = 0; i < count; i++)
    {
        printf("Valor de i: %d\n", i);
    }
    
    int i = 0;
    while (i < count)
    {
        printf("Valor de i: %d\n", i);
        i++;
    }
    

    int i = 0;
    do
    {
        printf("Valor de i: %d\n", i);
        i++;
    } while (i < count);
    
}



// ################################# Funciones #####################################################
// estructura: <tipo de dato> nombre (parametros) { ... }

void funcionEjemplo()
{
    printf("Esto es un print dentro de una funcion");
    char x = "Esto es una palabra dentro de una funcion";
}

int funcionEjemploSuma(int x, int y) //al ser una funcion de tipo distinto a void, tiene que devolver una valor
{
    int z = x + y;
    return z; //Si o si tiene que devolver un entero (o una funcion que devuelva un entero)
}

int devuelveUno(){return 1;}

int funcionReturnFuncion()
{
    return devuelveUno();
}



// ################################# Funcion Main ##################################################
// estructura: int main() { ... }

/**
 * return  0 = indica que el programa terminó correctamente. 
 * return  1 = indica que hubo un error. 
 * return -1 = indica que hubo un error grave.
*/

int main() {

    // Imprimir en consola
    printf("Hola, Mundo!\n");

    // Llamar a una funcion
    funcionEjemplo();

    int resultado = funcionEjemploSuma(5, 10);
    printf("\nEl resultado de la suma es: %d\n", resultado);

    return 0;
}



// ################################# Punteros #######################################################
// variable que contiene la direccion de otra variable

void explicacionPunteros()
{
    double a[4] = {1.1,2.2,3.3,4.4};
    double *z;      //Si no se le asigna nada, es puntero nulo
    double **w;     //puntero a un puntero (el ultimo puntero es el que tiene la direccion de memoria del dato)
    double *p,x;    //En la declaración se añade * antes del nombre de la variable
    p = &a[2];      //El operador & devuelve la dirección de una variable
    x = *p;
    *p = 0.0;
    p = a; /* &a[0] */
}



// ################################# Hilos ##########################################################

#include <pthread.h> //libreria necesaria para crear hilos
void funcionDeHilos()
{
    printf("hola mundo");
}
void Hilos()
{
    int data = 0;
    pthread_t hilo; //crea una entidad "hilo"

    pthread_create(&hilo, NULL, &funcionDeHilos, &data); //con esto creamos los hilos (sus parametros), TODOS ejecutan "funcionDeHilos", la variable data es compartida por los hilos

    pthread_join(hilo, NULL); //el hilo principal espera a que terminen antes de continuar con la siguiente instrucción del main
}



// ################################# Fork ##########################################################
#include <unistd.h> //libreria necesaria para usar fork
void funcionDeFork()
{
    printf("hola mundo");
}
void Fork()
{
    /**
     * El proceso padre y el proceso hijo ejecutan el mismo código a partir de la llamada a fork(), pero con un PID diferente.
     * El proceso padre recibe el PID del proceso hijo como valor de retorno de fork(), mientras
     * que el proceso hijo recibe 0 como valor de retorno de fork(). Esto permite que ambos procesos puedan ejecutar código diferente dependiendo de su rol (padre o hijo) después de la llamada a fork().
     * 
     * Ejemplo:
    */
    pid_t pid = fork(); //crea un proceso hijo, el proceso padre y el proceso hijo ejecutan el mismo codigo a partir de esta linea, pero con un PID diferente (pid del proceso padre es el PID del proceso hijo, y el PID del proceso hijo es 0)
    if (pid == 0) {
        // Código del proceso hijo
        funcionDeFork();
    } else if (pid > 0) {
        // Código del proceso padre
        funcionDeFork();
    } else {
        // Error al crear el proceso
        perror("Error al crear el proceso");
    }
}